### PR TITLE
8274753: ZGC: SEGV in MetaspaceShared::link_shared_classes

### PR DESCRIPTION
**Pull request by tstuefe**
Hi all,

not a clean backport. I had to fix up classLoaderData.cpp a bit because `SystemDictionaryShared::handle_class_unloading(ik);` does not yet exist in jdk17 and I did not want to bring the dependent cleanup change into jdk17 too (among other things because Oracle seems not to have done it either in their closed backport as far as I can glean from the JBS tags).

Tested in SAP nightlies for about a week now.

Thanks!

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issues
 * [JDK-8274753](https://bugs.openjdk.java.net/browse/JDK-8274753): ZGC: SEGV in MetaspaceShared::link_shared_classes
 * [JDK-8274935](https://bugs.openjdk.java.net/browse/JDK-8274935): dumptime_table has stale entry

### Reviewers
 * [Goetz Lindenmaier](https://openjdk.java.net/census#goetz) (@GoeLin - **Reviewer**)
 * [Paul Hohensee](https://openjdk.java.net/census#phh) (@phohensee - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk17u-dev pull/118/head:pull/118` \
`$ git checkout pull/118`

Update a local copy of the PR: \
`$ git checkout pull/118` \
`$ git pull https://git.openjdk.java.net/jdk17u-dev pull/118/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 118`

View PR using the GUI difftool: \
`$ git pr show -t 118`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk17u-dev/pull/118.diff">https://git.openjdk.java.net/jdk17u-dev/pull/118.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/classfile/classLoaderData.cpp [hunk 1]
```diff
@@ -886,6 +887,10 @@ void ClassLoaderData::free_deallocate_list_C_heap_structures() {
       // Remove the class so unloading events aren't triggered for
       // this class (scratch or error class) in do_unloading().
       remove_class(ik);
+      // But still have to remove it from the dumptime_table.
+      if (Arguments::is_dumping_archive()) {
+        SystemDictionaryShared::remove_dumptime_info(ik);
+      }
```
Comment (GoeLin) [inline src/hotspot/share/classfile/classLoaderData.cpp:893]: This is a part of handle_class_unloading(). I.e., you basically "undo" '8267189: Remove duplicated unregistered classes from dynamic archive'
which replaced remove_dumptime_info() by handle_class_unloading().
And you add the check for is_dumping_archive() that was 
added to handle_class_unloading() in 
'8265602: -XX:DumpLoadedClassList should support custom loaders'.

LGTM

**Review by GoeLin — APPROVED** [review]

**Review by phohensee — APPROVED** [review]
Lgtm.

**Comment by tstuefe** [comment]
@goelin, @phohensee thanks!
@GoeLin could you approve the fix request then?

File: src/hotspot/share/classfile/classLoaderData.cpp (hunk 1 above)
Comment (rgithubli) [reply src/hotspot/share/classfile/classLoaderData.cpp:893]: @GoeLin I'm backporting https://bugs.openjdk.org/browse/JDK-8267189 to 17 and found this comment. Having trouble understanding it. Is it still safe / recommended to backport JDK-8267189 with this unclean 8274753 backport? Thanks.

Comment (rgithubli) [reply src/hotspot/share/classfile/classLoaderData.cpp:893]: Attempted backport https://bugs.openjdk.org/browse/JDK-8267189, but the build failed due to these methods. I guess it won't be a clean backport. 


Not very familiar with the code, would you be able to give suggestion on where I should update? With JDK-8267189, can we remove `SystemDictionaryShared::remove_dumptime_info(ik);`?

```
=== Output from failing command(s) repeated here ===
* For target hotspot_variant-server_libjvm_objs_classLoaderData.o:
src/hotspot/share/classfile/classLoaderData.cpp: In member function 'void ClassLoaderData::free_deallocate_list_C_heap_structures()':
src/hotspot/share/classfile/classLoaderData.cpp:892:56: error: 'static void SystemDictionaryShared::remove_dumptime_info(InstanceKlass*)' is private within this context
         SystemDictionaryShared::remove_dumptime_info(ik);
                                                        ^
In file included from src/hotspot/share/classfile/classLoaderData.cpp:58:0:
src/hotspot/share/classfile/systemDictionaryShared.hpp:234:15: note: declared private here
   static void remove_dumptime_info(InstanceKlass* k) NOT_CDS_RETURN;
               ^~~~~~~~~~~~~~~~~~~~
At global scope:
cc1plus: error: unrecognized command line option '-Wno-cast-function-type' [-Werror]
cc1plus: all warnings being treated as errors

* All command lines available in /workplace/ruiamzn/github/jdk17u-dev/build/linux-x86_64-server-release/make-support/failure-logs.
=== End of repeated output ===

```

Comment (rgithubli) [reply src/hotspot/share/classfile/classLoaderData.cpp:893]: Looked through the code, [JDK-8267189](https://bugs.openjdk.org/browse/JDK-8267189) introduces `handle_class_unloading`.

It seems undoing this change `necessary changes for jdk17` as part of backporting `JDK-8267189` makes sense. (?) Tested locally with simply `make images`, can build.

Comment (rgithubli) [reply src/hotspot/share/classfile/classLoaderData.cpp:893]: Tested jtreg_tier1 with [current draft](https://github.com/openjdk/jdk17u-dev/pull/819/commits), failed at [TestUnloadClassError.java](https://github.com/openjdk/jdk17u-dev/blob/master/test/hotspot/jtreg/runtime/BadObjectClass/TestUnloadClassError.java)

Comment (GoeLin) [reply src/hotspot/share/classfile/classLoaderData.cpp:893]: Please don't comment in this old PR. Use the new one you opened.

**Pull request closed without merging**